Your first booster app in 10 minutes (#151)
boostercloud/booster: docs/README.md

- [First command](#2-first-command)
- [First event](#3-first-event)
- [First entity](#4-first-entity)
- [First read model](#5-first-read-model)
- [Deployment](#6-deployment)
- [Testing](#7-testing)
  - [Creating posts](#71-creating-posts)
  - [Retrieving all posts](#72-retrieving-all-posts)
  - [Retrieving specific post](#73-retrieving-specific-post)
- [Removing stack](#8-removing-stack)
- [More functionalities](#9-more-functionalities)

#### 1. Create the project

First of all, we will use the Booster CLI to create a project. Run this command and follow
the instructions, when asked for the provider, select AWS as that is what we have
configured [here](#set-up-an-aws-account)

```shell
> boost new:project boosted-blog

...

ℹ boost new 🚧
✔ Creating project root
✔ Generating config files
✔ Installing dependencies
ℹ Project generated!
```

> Booster's new: commands follows this structure
>
> - `boost` is the Booster CLI
> - `new:<resource>` new is a CLI command, :project tells booster the kind of resource
> - `boosted-blog` is a parameter for the `new:project` command
> project name

The `new:project` command generates some scaffolding for you. The project name will be the

> project's root so `cd` to it:

Reviewer: ```suggestion
project's root so `cd` into it:
```

```shell
cd boosted-blog
```

There you should have these files and directories already generated:

```text
boosted-blog
├── .eslintignore
├── .gitignore
├── package-lock.json
├── package.json
├── src
│   ├── commands
│   ├── common
│   ├── config
│   │   └── config.ts
│   ├── entities
│   ├── events
│   └── index.ts
└── tsconfig.json
```

Now open the project in your favorite editor, e.g. [Visual Studio Code](https://code.visualstudio.com/).

#### 2. First command

Commands define the input to our system, so we'll start by generating our first [command](#commands) to create posts. Use the command generator in the project's root directory as follows:

```bash
boost new:command CreatePost --fields postId:UUID title:string content:string author:string
```

The `new:command` generator creates a `CreatePost.ts` file in the `commands` folder:

```text
boosted-blog
└── src
    └── commands
        └── CreatePost.ts
```

We still need to define two more things:

1. Who is authorized to run this command.

> 1. And what events are being triggered when the command is executed.

Reviewer: ```suggestion
2. And what events are being triggered when the command is executed.
```

Author: For MD it doesn't matter what number you use because the parser will handle the ordered list automatically as soon as they start with `<number>.`

https://www.markdownguide.org/basic-syntax#ordered-lists

I think it is actually a good practice to just use `1.` because if you move the item you don't have to care about the new order

```

We are close to testing our app, and to do so, we will use the GraphQL API endpoint.
Let's address that in the next section.

#### 7. Testing

Let's get started testing the project. We will perform three actions:

- Add a couple of posts
- Retrieve all posts
- Retrieve a specific post

Booster applications provide you with a GraphQL API out of the box. Commands are mapped
to mutations and read models to queries. To perform calls to the GraphQL API, you can use
any HTTP client you want; we recommend you to use [Postwoman](https://postwoman.io/graphql).

##### 7.1 Creating posts

Use your favorite GraphQL client to run these mutations. No authorization header is
required since we have allowed `all` to execute our commands and query the read models.

```graphql
mutation {
  CreatePost(
    input: {
      postId: "95ddb544-4a60-439f-a0e4-c57e806f2f6e"
      title: "Build a blog in 10 minutes with Booster"
      content: "I am so excited to write my first post"
      author: "Boosted developer"
    }
  )
}
```

```graphql
mutation {
  CreatePost(
    input: {
      postId: "05670e55-fd31-490e-b585-3a0096db0412"
      title: "Booster framework rocks"

>       content: "I am so excited for writting the second post"

Reviewer: ```suggestion
      content: "I am so excited for writing the second post"
```

```shell
> boost new:project boosted-blog

...

ℹ boost new 🚧
✔ Creating project root
✔ Generating config files
✔ Installing dependencies
ℹ Project generated!
```

> Booster's new: commands follows this structure
>
> - `boost` is the Booster CLI
> - `new:<resource>` new is a CLI command, :project tells booster the kind of resource
> - `boosted-blog` is a parameter for the `new:project` command
> project name

The `new:project` command generates some scaffolding for you. The project name will be the
project's root so `cd` to it:

```shell
cd boosted-blog
```

There you should have these files and directories already generated:

```text
boosted-blog
├── .eslintignore
├── .gitignore
├── package-lock.json
├── package.json
├── src
│   ├── commands
│   ├── common
│   ├── config
│   │   └── config.ts
│   ├── entities
│   ├── events
│   └── index.ts
└── tsconfig.json
```

Now open the project in your most preferred IDE, e.g. [Visual Studio Code](https://code.visualstudio.com/).

#### 2. First command

We will now define our first [command](#commands), it will allow us to create posts in our
blog. Write this command in the project's root directory:

```bash

> boost new:command CreatePost --fields postId:UUID title:string content:string author:string

Reviewer: Perhaps a bank account app?

```

The `new:command` creates most of the code for us, take a look at it here:

```text
boosted-blog
└── src
    └── commands
        └── CreatePost.ts
```

We still need to define two more things:

1. Who is authorized to run this command.
1. And what events are being triggered when the command is executed.

We will create the first Event later on this guide so, let's start authorizing `'all'` to
run this command, to do so edit `src/commands/CreatePost.ts` like this:

```typescript
@Command({
  authorize: 'all', // Specify authorized roles here. Use 'all' to authorize anyone
})
export class CreatePost {
  public constructor(
    readonly postId: UUID,
    readonly title: string,
    readonly content: string,
    readonly author: string
  ) {}

  public handle(register: Register): void {
    register.events(/* YOUR EVENT HERE */)
  }
}

```

#### 3. First event

It makes sense to emit an event after executing the command `CreatePost`, something like

> `PostCreated`. Let's do it by running this command:

Author: I'll rewrite this suggestion to make it a bit simpler. I think it is not necessary to compare the approaches (table vs. event-sourced) at this point. Also, we are not updating the posts in the tutorial, so the `Later, when we edit the post...` part is not covered.

> `PostCreated`. Let's do it by running this command:

Reviewer: Yep, this is "a taste" of Booster, i'd make it very simple. In the architecture section the user can read more

```bash
boost new:event PostCreated --fields postId:UUID title:string content:string author:string
```

The `new:event` command generates a new file under the `src/events` directory. The name of
the file is the name of the Event:

```text
boosted-blog
└── src
    └── events
        └── CreatePost.ts
```

Edit the `entityID` method to return the id of the entity this Event is referring to. It
should look like this:

```typescript
// src/events/PostCreated.ts

@Event
export class PostCreated {
  public constructor(
    readonly postId: UUID,
    readonly title: string,
    readonly content: string,
    readonly author: string
  ) {}

  public entityID(): UUID {
    return this.postId;
  }
}
```

> Now that we have an event, we can make the `CreatePost` command to emit it. Let's change

Reviewer 2: I'm not sure about the use of `make` here. What do you think?

Author: I'm not really sure because [make is super-polysemic](https://www.merriam-webster.com/dictionary/make). In that sentence it sounds to me like _now we can make this thing to happen_ but I'll change it by `edit` because it keeps the sense and we avoid using make.

the command's `handle` method to look like this:

```typescript
// src/commands/CreatePost.ts::handle
public handle(register: Register): void {
  register.events(new PostCreated(this.postId, this.title, this.content, this.author))
}
```

#### 4. First entity

Now we have a command and an event, however, we do not have any data representation of a
`Post`. To do so, we will create an `entity`.

```bash
boost new:entity Post --fields title:string content:string author:string --reduces PostCreated
```

This time Booster has created a file called `Post.ts` in the `src/entities` directory.
An entity is a reduction function applied to a series of events. For this example, we won't
take into account the previous state of the `Post`, so we will create a new one

> from the Event's data.

Reviewer 2: Although it is not directly related with this PR, do we have in mind create other examples using the previous states?  I think that they will help a lot especially for more complex use cases.

Author: Thank you @rdiaz82 for pointing this out. Yep, we have already captured an examples section 👍

```typescript
// src/entities/Posts.ts
@Entity
export class Post {
  public constructor(public id: UUID, readonly title: string, readonly content: string, readonly author: string) {}

  @Reduces(PostCreated)
  public static reducePostCreated(event: PostCreated, currentPost?: Post): Post {
    return new Post(event.postId, event.title, event.content, event.author)
  }
}
```

#### 5. First read model

> So far, we have been adding data to our blog. Now, we will retrieve our data using read
> models. A read model allows us to query our entire entities or just a subset of their
> attributes. Let's project our `Post` Entity:

Reviewer 3: ```suggestion
In a real application, we rarely want to publish our entire domain model (entities) including all their fields, and we need to define which roles can access which subset of the data. We use ReadModels for that, to define the resources that client applications can query or subscribe to. We build our Read Models by projecting entities, let's project our `Post` Entity to produce a `PostReadModel`:
```

```bash
boost new:read-model PostReadModel --fields title:string author:string --projects Post:id
```

As you might guess, the command will generate a file called `PostReadModel.ts` under
`src/read-models`:

```text
boosted-blog
└── src
    └── read-models
        └── PostReadModel.ts
```

There are two things to do when creating a read model:

1. Define who is authorized to query it
1. and filter out the entity fields that are not needed in the read model

> Read models and commands constitute the public API of a Booster application. With the
> `CreatePost` command, we authorized `all` to execute it. This time we will be doing the
> same for the read model. Furthermore, We will exclude the `content` field from the `Post`
> entity so it won't be visible when querying this read-model. To do so, edit the class to
> look like this:

Reviewer 3: ```suggestion
Read models and commands compound the public API of a Booster application. We allowed
anyone to submit the `CreatePost` command. We will allow anyone to query our read model too. 

Just for learning, we will show you how to exclude the `content` field from the `Post` entity so it won't be visible
when querying this read-model, you can add it later if you want to keep working in this application later. 

To authorize anyone to query the read model and exclude the `content` field, edit the class to look like this:
```

```typescript
// src/read-models/PostReadModel.ts
@ReadModel({
  authorize: 'all'// Specify authorized roles here. Use 'all' to authorize anyone
})
export class PostReadModel {
  public constructor(
    public id: UUID,
    readonly title: string,
    readonly author: string,
  ) {}

  @Projects(Post, "postId")
  public static projectPost(entity: Post, currentPostReadModel?: PostReadModel): PostReadModel {
      return new PostReadModel(entity.id, entity.title, entity.author);
  }

}
```

#### 6. Deployment

> We have everything we need to deploy our application to the cloud. It is as simple as
> running this command:

Reviewer 3: ```suggestion
At this point, we've learned how to create a publicly accessible command, emit an event to store the data, declare how the event is reduced into an entity, that represents our system internal state, and projected it into a read model that is publicly accessible, so you already know the basics to build event-driven applications with Booster. 

Let's deploy our application to the cloud to see it working. It is as simple as running the deploy command:
```

```bash

> boost deploy -e production

Reviewer 2: what does mean the environment? Are there any differences if I deploy the booster app using development, testing or whatever value in that option?

Author: Excellent questions, the answer is I don't know :dumb: but I wouldn't explain the meaning and options in this 10m guide. Those will appear in the deploying section, and surely every provider will also need an explanation of what is the difference in using every environment.

Maybe @alvaroloes or @tainguyenbui can answer this better.

Reviewer 3: Booster allows defining more than one environment, similar to other frameworks like Rails. We introduced this feature when we started working in the local runtime, to be able to run the project locally using the `development` environment or publish it to the cloud using the `production` environment. There are no limitations on how you use this feature, so you could also configure staging environments or have personal environments for the developers using different accounts or parameterizing the app name.

```

It will take a couple of minutes to deploy all the resources. Once finished, you will see
information about your stack endpoints. For this example, we will only need to pick the
API endpoint. Look into the output for something like `backend-application-stack.baseRESTURL`
e.g.

```text
https://<some random number>.execute-api.us-east-1.amazonaws.com/production/graphql
```

We are close to testing our app, and to do so, we will use the GraphQL API endpoint.
Let's address that in the next section.

#### 7. Testing

Let's get started testing the project. We will perform three actions:

- Add a couple of posts
- Retrieve all posts
- Retrieve a specific post

Booster applications provide you with a GraphQL API out of the box. Commands are mapped
to mutations and read models to queries. To perform calls to the GraphQL API, you can use

> any HTTP client you want; we recommend you to use [Postwoman](https://postwoman.io/graphql).

Reviewer 3: ```suggestion
any HTTP client you want; we recommend you to use [Postwoman](https://postwoman.io/graphql), which is free and includes great support for GraphQL.
```

##### 7.1 Creating posts

Use your favorite GraphQL client to run these mutations. No authorization header is
required since we have allowed `all` to execute our commands and query the read models.

```graphql
mutation {
  CreatePost(
    input: {
      postId: "95ddb544-4a60-439f-a0e4-c57e806f2f6e"
      title: "Build a blog in 10 minutes with Booster"
      content: "I am so excited to write my first post"
      author: "Boosted developer"
    }
  )
}
```

```graphql
mutation {
  CreatePost(
    input: {
      postId: "05670e55-fd31-490e-b585-3a0096db0412"
      title: "Booster framework rocks"
      content: "I am so excited for writting the second post"
      author: "Another boosted developer"
    }
  )
}
```

The expected response for each of those requests should be:

```json
{
  "data": {
    "CreatePost": true
  }
}
```

Note:

> > The IDs are generated on the client-side. For this example, we are not validating ID

Reviewer 2: I don't know why but the first time that I read the sentence, it seems to me that there is also the possibility to configure the server-side ID generation

Author: You are right @rdiaz82, thank you for pointing this out. I'll rephrase this to make the point clearer

Author: what do you think about this:

> With Booster, IDs are generated on the client-side. When running production applications
> consider adding validation for ID uniqueness. In this example, we have used [a UUID generator](https://www.uuidgenerator.net/version4)

Reviewer 3: Well, this will change soon, we've planned to return the `entityID`s of all events emitted, instead of just that unmeaningful "true"

> uniqueness, consider using [a UUID generator](https://www.uuidgenerator.net/version4)
> when creating more Posts.

##### 7.2 Retrieving all posts

Let's perform a GraphQL `query` that will be hitting our `PostReadModel`:

```graphql
query {
  PostReadModels {
    id
    title
    author
  }
}
```

> It should response something like:

Reviewer 3: ```suggestion
It should respond with something like:
```

```json
{
  "data": {
    "PostReadModels": [
      {
        "id": "05670e55-fd31-490e-b585-3a0096db0412",
        "title": "Booster framework rocks",
        "author": "Another boosted developer"
      },
      {
        "id": "95ddb544-4a60-439f-a0e4-c57e806f2f6e",
        "title": "Build a blog in 10 minutes with Booster",
        "author": "Boosted developer"
      }
    ]
  }
}
```

##### 7.3 Retrieving specific post

It is also possible to retrieve specific a `Post` by adding the `id` as input, e.g.:

```graphql
query {
  PostReadModel(id: "95ddb544-4a60-439f-a0e4-c57e806f2f6e") {
    id
    title
    author
  }
}
```

You should get a response similar to this:

```json
{
  "data": {
    "PostReadModel": {
      "id": "95ddb544-4a60-439f-a0e4-c57e806f2f6e",
      "title": "Build a blog in 10 minutes with Booster",
      "author": "Boosted developer"
    }
  }
}
```

> #### 8. Removing stack

Reviewer 3: ```suggestion
#### 8. Removing the stack
```

> Now, let's destroy all the infrastructure created for running the application in the cloud.

Reviewer 3: ```suggestion
It is convenient to destroy all the infrastructure created after you stop using it to avoid generating cloud resource costs.
```

> Execute the following command from the root of the project. For security reasons, you have

Reviewer 3: I think that security implies other more dramatic meanings, like avoiding data leaks or something like that, safety sounds softer to me...

```suggestion
Execute the following command from the root of the project. For safety reasons, you have
```

to confirm this action by writing the project's name, in our case `boosted-blog` that is
the same used when we run `new:project` CLI command.

```bash
> boost nuke -e production

? Please, enter the app name to confirm deletion of all resources: boosted-blog
```

Note:
> Congratulations! You've built a serverless backend in less than 10 minutes. We hope you
> have enjoyed discovering the magic of the Booster Framework.

#### 9. More functionalities

The are many other options for your serverless backend built with Booster Framework:

> - Authorize commands and read models based on different roles
> - Use GraphQL subscriptions
> - Make events to trigger other events
> - Serving static content
> - Reading entities within command handlers to apply domain-driven decisions
> - and much more...

Reviewer 3: ```suggestion
- Build more complex authorization schemas for commands and read models based on user roles
- Use GraphQL subscriptions to get updates in real-time
- Make events trigger event handlers that can run code or trigger other events
- Deploy static content
- and much more...
```

We're still discussing this => Reading entities within command handlers to apply domain-driven decisions

Reviewer: I'd rewrite this part in your suggestion

```diff
- - Make events trigger event handlers that can run code or trigger other events
+ - Make events trigger code or other events
```

Author: I finally wrote: `- Make events trigger other events`

> Please continue reading to get more details about Booster and its capabilities.

Reviewer 3: ```suggestion
Continue reading to dig more, you've just scratched the surface of all the Booster capabilities!
```
